book/src/dev/rfcs/0003-state-updates.md

```diff
@@ -0,0 +1,305 @@
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
+
+Zcash structures are encoded using `ZcashSerialize`/`ZcashDeserialize`.
+
+## Request / Response API
+[request-response]: #request-response
+
+The state API is provided by a pair of `Request`/`Response` enums. Each
+`Request` variant corresponds to particular `Response` variants, and it's
+fine (and encouraged) for caller code to unwrap the expected variants with
+`unreachable!` on the unexpected variants. This is slightly inconvenient but
+it means that we have a unified state interface with unified backpressure.
+
+This API includes both write and read calls. Spotting `Commit` requests in
+code review should not be a problem, but in the future, if we need to
+restrict access to write calls, we could implement a wrapper service that
+rejects these, and export "read" and "write" frontends to the same inner service.
+
+### `Request::CommitBlock(Arc<Block>)`
+[request-commit-block]: #request-commit-block
+
+Performs contextual validation of the given block, committing it to the state
+if successful. Returns `Response::Added(BlockHeaderHash)` with the hash of
+the newly committed block or an error.
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.
+
+Otherwise, attempt to perform contextual validation checks and the commit the
+given block to the state. The exact list of contextual validation checks will
+be specified in a later RFC. If contextual validation checks succeed, the new
+block is added to one of the in-memory chains. If the resulting chain is
+longer than 100 blocks, the oldest block is now past the reorg limit, so we
+remove it from all in-memory chains and commit it to sled as described below
+in `CommitFinalizedBlock`.
+
+Finally, process any queued children of the newly committed block the same way.
+
+### `Request::CommitFinalizedBlock(Arc<Block>)`
+[request-commit-finalized-block]: #request-finalized-block
+
+Commits a finalized block to the sled state, skipping contextual validation.
+The block's parent must be the current sled tip. This is exposed for use in
+checkpointing, which produces in-order finalized blocks. Returns
+`Response::Added(BlockHeaderHash)` with the hash of the committed block if
+successful.
+
+This should be implemented as a wrapper around a function also called by
+[`Request::CommitBlock`](#request-commit-block), which should:
+
+1. Obtain the highest entry of `hash_by_height` as `(old_height, old_tip)`.
+Check that `block`'s parent hash is `old_tip` and its height is
+`old_height+1`, or panic.  This check is performed as defense-in-depth
+to prevent database corruption, but it is the caller's responsibility to
+commit finalized blocks in order.
```

Comment:
Here's the core issue in this part of the RFC:

I think option 1 assigns the chain order constraint to the module that is best able to deal with it. It's also the easiest check to code correctly, because it can be done synchronously. (So race conditions or deadlocks should be impossible.) But it does introduce yet another cache of "not-quite-ready" blocks, so I can understand if you'd prefer one of the other options.

But the RFC makes the caller is responsible, which implies option 2 or 3. They seem harder to get right - we've already seen deadlocks between the sync service and checkpoint verifier.

Reply:
Hmm, I see. Yes, this is a problem.  The motivation for this constraint was exactly the one that you described, wanting to avoid adding another cache of not-quite-ready blocks.

I wonder whether this problem is downstream of the way that we've set up the relations between the verifiers and the state.  I don't think that the `ChainVerifier` should be committing blocks, because the mechanism for committing blocks is different depending on the verification method (checkpoint or full verification).  If the `ChainVerifier` committed any blocks, we'd have to have separate logic for each path, but at that point there's no common behaviour to put in the `ChainVerifier`, so it would be simpler to have each verifier be responsible for committing verified blocks, and for the `ChainVerifier` to be a very lightweight wrapper that dispatches requests to one or the other.

After choosing that division of responsibility, there's no issue with the block verifier, because the `CommitBlock` request is allowed to be performed out-of-order. So the issue is only with how to commit checkpointed blocks in order, but this is easy because at the point that we do the verification, we have all the blocks in order.   T

Reply:
That sounds like a good plan, I will resolve #969 by moving the AddBlock calls to the block and chain verifiers. Then we can replace them with the appropriate new calls as needed.

(The BlockVerifier currently implements a workaround to make sure blocks are in order, so we will resolve the bug by making this change.)